internal/block/block.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package block
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+// Block represents the buffer which holds the data.
+type Block interface {
+	// Reuse resets the blocks for reuse.
+	Reuse()
+
+	// Size provides the current data size of the block. The capacity of the block
+	// can be >= data_size.
+	Size() int64
+
+	// Write writes the given data to block.
+	Write(bytes []byte) error
+
+	// Reader interface helps in copying the data directly to storage.writer
+	// while uploading to GCS.
+	Reader() io.Reader
+}
+
+// TODO: check if we need offset or just storing end is sufficient. We might need
+// for handling ordered writes. It will be decided after ordered writes design.
+type offset struct {
+	start, end int64
+}
+
+type memoryBlock struct {
+	Block
+	buffer []byte
+	offset offset
+}
+
+func (m *memoryBlock) Reuse() {
```

Comment:
Eventually we would need Reuse function to take a `Block chan` as parameter and freed up `memoryBlock` will be put in that Block chan for reuse. Shall we do this in subsequent PR when required?

Reply:
I am planning to make it the responsibility of the caller to put the block after reuse to the channel (basically what to do after reuse). The block struct shoudn't worry about it

Reply:
SG, thanks for clarifying!